KMS DSI 7" panel support

6by9 opened the pull request:
This gets the DSI panel working on Pi3. Pi4 is still a work in progress.

The touchscreen is using the upstream driver for the TS controller chip, but that mandates that the interrupt line be wired up, and it isn't on our panel.
For some reason that driver can't detect the TS controller if loaded during boot (ie ```dtoverlay=edt-ft5406``` in config.txt), but is fine from the command line. The reason is not understood at present.

(I'm just creating the matching PR for 5.10 as well).

6by9 committed:
drm/panel/raspberrypi-touchscreen: Use independent I2C actions with delay.

We now have the hardware I2C controller pinmuxed to the drive the
display I2C, but this controller does not support clock stretching.
The Atmel micro-controller in the panel requires clock stretching
to allow it to prepare any data to be read.

Split the rpi_touchscreen_i2c_read into two independent transactions with
a delay between them for the Atmel to prepare the data.

Signed-off-by: Dave Stevenson <dave.stevenson@raspberrypi.com>

6by9 committed:
Input: edt-ft5x06: Poll the device if no interrupt is configured.

Not all systems have the interrupt line wired up, so switch to
polling the touchscreen off a timer if no interrupt line is
configured.

Signed-off-by: Dave Stevenson <dave.stevenson@raspberrypi.com>

6by9 committed:
dtoverlays: Add an overlay for the EDT FT5406 touchscreen

This touchscreen controller is used by the 7" DSI panel, and
this overlay configures it for when it is NOT being polled by
the firmware.

Signed-off-by: Dave Stevenson <dave.stevenson@raspberrypi.com>

6by9 committed:
drm/panel/raspberrypi-ts: Insert delay before polling for startup state

In switching to the hardware I2C controller there is an issue
where we seem to not get back the correct state from the Pi
touchscreen.
Insert a delay before polling to avoid this condition.

Signed-off-by: Dave Stevenson <dave.stevenson@raspberrypi.com>

6by9 committed:
Add devicetree support for RaspberryPi 7" panel over DSI I2C

Signed-off-by: Dave Stevenson <dave.stevenson@raspberrypi.com>

6by9 committed:
drm/vc4: Add support for DSI0

DSI0 was partially supported, but didn't register with the main
driver, and the code was inconsistent as to whether it checked
port == 0 or port == 1.

Add compatible string and other support to make it consistent.

Signed-off-by: Dave Stevenson <dave.stevenson@raspberrypi.com>

6by9 commented:
I think I now understand the touchpanel dtoverlay issue.

The Atmel on the daughterboard includes a line to the touchscreen reset line. Default is to hold the TS in reset, and the REG_POWERON command will release that reset.

The touchscreen driver knows nothing about the Atmel, therefore it's a race as to which gets initialised first. Generally the TS is first and hence the TS is still in reset.
Time to add a load more dependencies :-/ I seem to recall there was a thread on dri-devel recently that was looking at something around this lot.

6by9 commented:
Found. It was adding it as a regulator controller, which will probably do the trick.

They're in 5.10
https://github.com/torvalds/linux/commit/9c535960891470eaec5664eed2a56dbb2b6a205b
https://github.com/torvalds/linux/commit/38573472278da49e2ab588062d88bc2f0c11a325


There's also patches to make the TC358762 an independent driver, which are in 5.10.
https://github.com/torvalds/linux/commit/1930d19cb49930b7d5be0379be9f88fcbeba9976
https://github.com/torvalds/linux/commit/97722439eeed2f2e4c1a1d862faa0ee7bc2783c3

Now to rework it all to see if we can get the DT side to match up!
I think that'll be a separate PR to use the newer drivers, as this is doing a reasonable job at present.

pelwell commented:
For obvious reasons this has the same review feedback as #3955

vianpl commented on drivers/input/touchscreen/edt-ft5x06.c:
```diff
@@ -1221,17 +1242,27 @@ static int edt_ft5x06_ts_probe(struct i2c_client *client,
+		timer_setup(&tsdata->timer, edt_ft5x06_ts_irq_poll_timer, 0);
+		tsdata->timer.expires = jiffies +
+					msecs_to_jiffies(POLL_INTERVAL_MS);
+		add_timer(&tsdata->timer);
```
Not that important since it won't affect functionality, but would make the patch simpler. There is an API for polled input devices:

```
input_setup_polling()
input_set_poll_interval()
...
```

6by9 replied:
Does it work? #3777 is still under investigation for the firmware based TS driver where it isn't polling at the rate requested.

vianpl replied:
Sorry, I wasn't aware of that issue. It's indeed very bad with HZ=100 and only barely acceptable with HZ=250 (upstream arm64's default).  I guess a workaround would be to round down the jiffies count. I wonder if moving the implementation to hrtimers when the polling delay has the same order of magnitude as HZ's delta makes sense.

vianpl commented:
@6by9 Any plans on upstreaming this series?

6by9 commented:
@vianpl 
> Any plans on upstreaming this series?

You generally know our approach in offloading the upstreaming to others.
Maxime is contracted to look after the KMS driver side, so we'll probably ask him to look at upstreaming it when it's all working. 
Getting Pi4 DSI working is now my main priority.

A chunk of this is also likely to be superceded by the TC358762 and Pi display regulator driver from 5.10, but one step at a time.

vianpl commented:
> You generally know our approach in offloading the upstreaming to others.

Of course, but this being a smaller driver I was unsure whether to add it to my TODO list. :)

> Maxime is contracted to look after the KMS driver side, so we'll probably ask him to look at upstreaming it when it's all working.
> Getting Pi4 DSI working is now my main priority.

That's perfect, thanks!

6by9 committed:
dt-bindings: Add compatible for BCM2711 DSI1

DSI1 on BCM2711 doesn't require the DMA workaround that is used
on BCM2835/6/7, therefore it needs a new compatible string.

Signed-off-by: Dave Stevenson <dave.stevenson@raspberrypi.com>

6by9 committed:
drm/vc4: Add configuration for BCM2711 DSI1.

BCM2711 DSI1 doesn't have the issue with the ARM not being
able to write to the registers, therefore remove the DMA
workaround for that compatible string.

Signed-off-by: Dave Stevenson <dave.stevenson@raspberrypi.com>

6by9 committed:
dt: Use compatible string for BCM2711 DSI1

Updates the compatible string for DSI1 on BCM2711 to
differentiate it from BCM2835.

Signed-off-by: Dave Stevenson <dave.stevenson@raspberrypi.com>

6by9 commented:
Updated to address Phil's review comments on #3955, and as Pi4 is now at least partially working.

Sending DSI commands seems to be slightly unreliable and sometimes times out, so the TC358762 isn't always configured fully. The same is true on Pi3, so I think the host_transfer part of the DSI driver needs a closer examination.

jamwu commented:
@6by9 Thanks for your great job! I am trying to used dsi1 on cm4. Which kernel should I choose? 5.9 or 5.10. Is the KMS driver ready for cm4 now?

pelwell commented:
5.10 is going to be our next "master" kernel - I suggest you try it and report any problems you encounter.

6by9 commented:
I was developing on 5.9, but all patches should now be in 5.10 as well. All future changes will be on 5.10.
I haven't tested on the CM4 as yet, but will be doing so in the next couple of days.

jamwu commented:
> 5.10 is going to be our next "master" kernel - I suggest you try it and report any problems you encounter.
Thanks for your advice. I will give it a try.

jamwu commented:
> I was developing on 5.9, but all patches should now be in 5.10 as well. All future changes will be on 5.10.
> I haven't tested on the CM4 as yet, but will be doing so in the next couple of days.

Understand. Hope to hear the good news about CM4 next month.
